Implement Fully-Featured Event Self-Registration for Images & Texts

Roland-J opened the pull request:
Adds the plumbing required for images to support image:register_event('left_click', function) and image:register_event('drag', function), potentially making images much more friendly to work with by lowering the technical barrier to utilizing click events for images. _(This functionality could also be extended to images if this PR is found desirable)_

Sample addon for demo testing:
![image](https://user-images.githubusercontent.com/107378114/178122859-37deaf08-8561-4e22-a062-8de974142a2f.png)

```
--------------------------------------------------------------------------------------------------------------
-- 1. Nonddraggable-image event self(:)-registration demo
--------------------------------------------------------------------------------------------------------------

-- Setup an image (nothing new here)
local images = require('images')
local image2 = images.new()
image2:pos(50, 75)
image2:size(175, 23)
image2:fit(false)
image2:path(windower.addon_path .. 'Button002.png')
image2:draggable(false)
image2:show()

-- Self(:)-register a left-click event (NEW!)
image2:register_event('left_click', function(...)
    print('I\m not draggable and I was just left_clicked!')
end)



--------------------------------------------------------------------------------------------------------------
-- 2. Draggable-image event self(:)-registration demo
-- NOTE: We only fire the left_click event on release and only if there was no dragging
--------------------------------------------------------------------------------------------------------------

local image = images.new()
image:pos(50, 50)
image:size(175, 23)
image:fit(false)
image:path(windower.addon_path .. 'Button002.png')
image:draggable(true)
image:show()

-- Self(:)-register a left-click event (NEW!)
image:register_event('left_click', function(...)
    print('I\'m draggable and I\'ve been clicked and unclicked, without being dragged!')
end)

-- Self(:)-register a drag event! (NEW!)
image:register_event('drag', function(...)
    local curr_x, curr_y = image:pos()
    local last_x, last_y = image:last_pos() --new function: provides the pos_x and pos_y of previous drag's final position
    local x_delta, y_delta = curr_x - last_x, curr_y - last_y --how to get current drag's precise x/y deltas, can be used to move clustered elements
    print('I\'ve been dragged and my delta is ' .. x_delta .. '/' .. y_delta .. '!')
end)
```

Byrth committed:
fix shortcuts (requires monster_skills.lua)

Byrth committed:
version bumps

Chiaia committed:
Merge pull request #2079 from funkworkz/patch-2

Update setbgm.lua

z16 committed:
Merge pull request #2080 from Windower/gearswap-shortcuts-monstrosity-fix

Gearswap shortcuts monstrosity fix

trv6 committed:
A few functions for decoding dialog entries

Locals, spacing


dev table, format arg, no more length


the rest of the owl

Chiaia committed:
Update messages and Add Low/Mid D Tower to be extracted

Chiaia committed:
Merge pull request #2083 from Windower/dev

Dev -> Live

z16 committed:
Merge pull request #2081 from trv6/dialog

A few functions for decoding dialog entries

Byrth committed:
change based on CHA CHA CHA's info

Byrth committed:
Merge pull request #2084 from Windower/adjust_0x01C_incoming

change based on CHA CHA CHA's info

Byrth committed:
fixed comment numbering for 0x01C

Byrth committed:
Merge pull request #2085 from Windower/fix_comment_numbering

fixed comment numbering for 0x01C

Chiaia committed:
Merge pull request #2087 from Windower/dev

Dev -> Live

zetaepyon committed:
Update README.md

xurion committed:
Update Mount Roulette to 3.1.0.
Add blacklist function.

RubenatorX committed:
Master Levels Update 1

alphaONE2 committed:
Merge pull request #2091 from RubenatorX/patch-12

Master Levels Update 1

KenshiDRK committed:
[thtracker] variable fix

variable fix

KenshiDRK committed:
Slips lib Update

Adding Chemistry Set and Goujian to Slips library

Nifim committed:
Merge pull request #2093 from KenshiDRK/Slips

Slips lib Update

Chiaia committed:
Merge pull request #2092 from KenshiDRK/Thtracker

[thtracker] variable fix

Chiaia committed:
Fix dupe label

Chiaia committed:
Merge pull request #2094 from Windower/dev

Dev -> Live

RubenatorX committed:
Update Master Level 2

Changed to use boolbits and added Master Levels menu display data

RubenatorX committed:
Master Levels Update 2

Change to use Refs

RubenatorX committed:
Master Level Update 2

Re-add Master flags comment

RubenatorX committed:
Master Level Update 2

spacing fix

z16 committed:
Merge pull request #2096 from RubenatorX/patch-15

Update Master Level 2

xurion committed:
Use set directly in settings table.
Use convenient iterator methods.
PR feedback.

Byrth committed:
the table's hex keys are interpreted as unsigned

RubenatorX committed:
Update Pointwatch.lua

1) Add support for Exemplar Points
2) Partial rework of packets to both utilize the packets library, and to take advantage of last_incoming when it is first loaded
3) Partial rework of Key Item detection for Dynamis

RubenatorX committed:
Update Pointwatch - Statics.lua - Exemplar Points

trv6 committed:
Use the other libs/dialog

RubenatorX committed:
Update Pointwatch.lua -- minor bugfix typo

RubenatorX committed:
update for 0x063 order 2

RubenatorX committed:
fix extra spaces

trv6 committed:
Swap the order of unpack's arguments.

trv6 committed:
Handle unsigned int

z16 committed:
Merge pull request #2098 from trv6/dialog

Swap the order of unpack's arguments.

paulframe committed:
Revert "update for 0x063 order 2"

This reverts commit f8015a0adb7e0bf777ebe8d5f27f3af257cb3842.

RubenatorX committed:
Update 0x063-Order 2

RubenatorX committed:
un-un-revert a change

RubenatorX committed:
fix broken fix

RubenatorX committed:
copy paste hates me

RubenatorX committed:
remove extra spaces

RubenatorX committed:
delete additional spaces

z16 committed:
Merge pull request #2099 from RubenatorX/patch-18

Packets Update 0x063-Order 2

KenshiDRK committed:
Update fields.lua

RubenatorX committed:
Add missing message ID

Byrth committed:
Add 0x0AD Incoming

Byrth committed:
Merge pull request #2100 from Windower/add-0x0AD-incoming

Add 0x0AD Incoming

z16 committed:
Merge pull request #2097 from RubenatorX/patch-17

Update Pointwatch.lua -- Exemplar Points

Chiaia committed:
Merge pull request #2101 from Windower/dev

Dev -> Live

z16 committed:
Merge pull request #2090 from xurion/dev

Update Mount Roulette to 3.1.0

KenshiDRK committed:
[fields.lua] Fix for Accolades

Fix for Accolades (Unity points)

Chiaia committed:
Merge pull request #2102 from KenshiDRK/dev

[fields.lua] Fix for Accolades

Chiaia committed:
Merge pull request #2103 from Windower/dev

Dev -> Live

Chiaia committed:
Box Destroyer Dec Update / Dressup + 2 items

Byrth committed:
add three items to slip 3 and one to slip 22

Byrth committed:
Merge pull request #2104 from Windower/update-slips-for-Dec-2021

add three items to slip 3 and one to slip 22

Byrth committed:
Merge pull request #2105 from Windower/dev

Dec 2021 update merge

z16 committed:
Libs/resources: Fixed lookup table name generation

trv6 committed:
Make it easy to get the dialog files.

trv6 committed:
Shadowing

Turn off shadows to reduce lag

trv6 committed:
Japanese dats

trv6 committed:
It would have been funny

trv6 committed:
Missed formatting

trv6 committed:
Prevent id overlap

They're pretty close!

trv6 committed:
Moved id conversion to dialog lib

trv6 committed:
Scrape the top two bits off the message ID

Chiaia committed:
Exemplar Points Documentation

Chiaia committed:
More modifiers

Byrth committed:
Update for 0x0B6 outgoing changes

Byrth committed:
Merge pull request #2109 from Windower/B6-outgoing-changes

Update for 0x0B6 outgoing changes

Byrth committed:
Merge pull request #2107 from trv6/dialog

Make it easy to get the dialog files.

Byrth committed:
Merge pull request #2108 from trv6/highbit

pointwatch: Scrape two high bits off the message ID so that it matches the dialog file ID

RubenatorX committed:
Correct improper ID range check (GearSwap)

RubenatorX committed:
remove unnecessary line and move lower out of loop

z16 committed:
Merge pull request #2110 from RubenatorX/patch-20

Correct improper ID range check

Chiaia committed:
Pilgrim/Nomad support/fix

Chiaia committed:
Merge branch 'dev' of https://github.com/Windower/Lua into dev

Chiaia committed:
Merge pull request #2046 from ianandersonlol/patch-9

Fix JobChange to work properly in Odyssey

Chiaia committed:
Merge pull request #2111 from Windower/dev

Dev -> Live

RubenatorX committed:
GearSwap Motes Command-- set mode with spaces

Allow to set a specific mode name with a space in it via `gs c set` (previously would ignore anything after a space, even if contained in quotes)

RubenatorX committed:
Gearswap Motes Command Fix

oops, this is the correct change, my bad

z16 committed:
Merge pull request #2112 from RubenatorX/patch-21

GearSwap Motes Command-- set mode with spaces

Byrth committed:
Index res.monster_abilities once

Byrth committed:
Also fix it in the actions library

Byrth committed:
Merge pull request #2114 from Windower/make-monster-abilities-non-lazy

eager-load resources for actions.lua

z16 committed:
Libs/actions: Adjusted to not use rawget for resources

Byrth committed:
Switch indinope to use the modified packet

Byrth committed:
Merge pull request #2115 from Windower/switch-indinope-to-modified

Switch indinope to use the modified packet

trv6 committed:
Verify logged in. Only update once.

Chiaia committed:
Jan 10th, 2022 Update

Chiaia committed:
Merge pull request #2116 from Windower/dev

Dev -> Live

KenshiDRK committed:
Add New items to Slips lib

Adding Feline Hagoita and Feline Hagoita +1

RubenatorX committed:
GearSwap SelfCommand handle_update fix

some luas apparently call this function directly so, have to convert to library table before calling slice

z16 committed:
Lib/tables: Forwarded table.sconcat parameters

z16 committed:
Merge pull request #2118 from RubenatorX/patch-22

GearSwap SelfCommand handle_update fix

Byrth committed:
Merge pull request #2117 from KenshiDRK/Slips

Add New items to Slips lib

Byrth committed:
Minor update to translate in 2022

Byrth committed:
Merge pull request #2119 from Windower/update-translate-2022

Minor update to translate in 2022

Chiaia committed:
Merge pull request #2120 from Windower/dev

Dev -> Live

Chiaia committed:
Formatting

Not saying I will ever touch this v4 version again but if I do I want to be able to read some of it.

Chiaia committed:
Merge branch 'dev' of https://github.com/Windower/Lua into dev

Byrth committed:
Add chocobo digging action enum

Byrth committed:
Merge pull request #2121 from Windower/add-chocobo-digging-action-enum

Add chocobo digging action enum

KenshiDRK committed:
[Battlemod] some fixes

fixed mobwscol and mobspellcol not having effect.
Also fixed items that give experience points and mweya plasm, now the simplified message after using them show what they give.

z16 committed:
Merge pull request #2122 from KenshiDRK/battlemod

[Battlemod] some fixes

KenshiDRK committed:
[ohShi] nil error fix

Some mobs 2hr use the monster abilities version of them so the param value is not on the job_abilities.lua.

Nifim committed:
Merge pull request #2123 from KenshiDRK/ohShi

[ohShi] nil error fix

RubenatorX committed:
Gearswap -- update usable and equippable bags

New Wardrobes

RubenatorX committed:
Organizer -- add wardrobes

RubenatorX committed:
FindAll -- Add Wardrobes

RubenatorX committed:
Itemizer -- Add Wardrobes

Byrth committed:
Add recycle to the ignore list

Byrth committed:
Merge pull request #2133 from Windower/fix-recycle-for-shortcuts

Add recycle to the ignore list

RubenatorX committed:
Packets 0x01C update

z16 committed:
Merge pull request #2128 from RubenatorX/patch-23

Gearswap -- update usable and equippable bags

z16 committed:
Merge pull request #2129 from RubenatorX/patch-24

Organizer -- add wardrobes

z16 committed:
Merge pull request #2130 from RubenatorX/patch-25

FindAll -- Add Wardrobes

z16 committed:
Merge pull request #2132 from RubenatorX/patch-27

Itemizer -- Add Wardrobes

KenshiDRK committed:
Packets update for new wardrobes

Update incoming 0x01C and 0x037 for the new wardrobes.

z16 committed:
Merge pull request #2138 from KenshiDRK/packet-lib

Packets update for new wardrobes

RubenatorX committed:
Packets -- Update 0x01D

Byrth committed:
Merge pull request #2139 from RubenatorX/patch-29

Packets -- Update 0x01D

RubenatorX committed:
Merge branch 'dev' into patch-28

RubenatorX committed:
Packets 0x01C Update

Byrth committed:
Merge pull request #2134 from RubenatorX/patch-28

Packets 0x01C update

Byrth committed:
Fix padding for 0x01C

Chiaia committed:
Merge pull request #2140 from Windower/dev

Dev -> Live

RubenatorX committed:
Organizer-Lib -- Add Wardrobes

Nifim committed:
Merge pull request #2141 from RubenatorX/patch-30

Organizer-Lib -- Add Wardrobes

Nifim committed:
Merge pull request #2142 from Windower/dev

Organizer-Lib -- Add Wardrobes from Windower/dev

RubenatorX committed:
Packets - 0x01D new boolbits and more info

z16 committed:
Merge pull request #2143 from RubenatorX/patch-33

Packets - 0x01D new boolbits and more info

z16 committed:
Merge pull request #2065 from KenshiDRK/packets-lib

Update fields.lua

funkworkz committed:
Update setbgm.lua

New song added in Feb update. We still do not know the names. Once we get an OST I will add names.

Nifim committed:
Merge pull request #2145 from funkworkz/patch-3

Update setbgm.lua

Byrth committed:
add new 0x01D handling to instaLS

Byrth committed:
Merge pull request #2146 from Windower/update-instals-for-0x01D

add new 0x01D handling to instaLS

Byrth committed:
Update instaLS.lua

Byrth committed:
fix both problems

Byrth committed:
bump version

Byrth committed:
Merge pull request #2149 from Windower/boxdestroyer-fails-to-account-for-monty-hall-and-is-too-verbose

BoxDestroyer refresh

Byrth committed:
Update README for changes

zetaepyon committed:
Update README.md

Escape on `<global>`

Byrth committed:
Merge pull request #2150 from Windower/dev

dev -> live 2022-02-14

Byrth committed:
Merge pull request #2151 from Windower/live

Live -> Dev 2022-02-14

KenshiDRK committed:
Slips lib update

Adding February items to the slips lib.

z16 committed:
Merge pull request #2153 from KenshiDRK/Slips

Slips lib update

Sechs committed:
Update validate.lua

Hi, I expanded the "Validate inventory" function. Normally it just works on a single repository (your inventory). I modified so you can do it on any repository, one at a time. Validate wardrobe3, validate safe, validate storage etc. It's updated with the newest additions of Wardrobes5-8.

If you use a custom Lua that includes all the Sets from all your jobs lua, you can basically check one repository at a time to find items that are present in those repositories but no longer present in any lua, i.e. items you no longer use.
It's basically a tool to clean up repositories from items you no longer use but are not aware of.
Multiple people have been using this tool but until now it required to temporarily overwrite the default validate with my custom one, then restore the original one after you're done with the cleaning.
Implemeting these changes in the basic Validate.lua would make this step no longer necessary.

More detailed info on how it works can be find in the related FFXIAH thread
https://www.ffxiah.com/forum/topic/54596/custom-gearswap-validate-to-clean-up-inventory/#3495138


There is one thing I'm unsure of.
I originally developed this custom lua several years ago, and I recently updated it adding Wardrobe5-8.
This means the original Validate.lua I used as a core for my custom might have been older than the current Validate.lua
Someone more competent than me would need to doublecheck that I didn't miss any important (new) feature in my custom version.
Shouldn't take long, I didn't really add too many lines to the original code in the end.

Thanks for your time.

Sechs committed:
Update validate.lua

Updated indentation for better reading

Sechs committed:
Update validate.lua

fixed indentation some more >_>

z16 committed:
Merge pull request #2155 from Sechs/patch-2

Update validate.lua

KenshiDRK committed:
[Battlemod] Fix for aoe ws skillchains

Fixing for aoe ws skillschains reporting all targets and not just the one that effectively got the skillchain.
Reported on issue https://github.com/Windower/Lua/issues/2154

z16 committed:
GearSwap: Minor adjustments to bag validation

z16 committed:
Merge pull request #2156 from KenshiDRK/battlemod

[Battlemod] Fix for aoe ws skillchains

RubenatorX committed:
Gearswap Validate syntax fix

zetaepyon committed:
Merge pull request #2157 from RubenatorX/patch-34

Gearswap Validate syntax fix

RubenatorX committed:
equipviewer validate bags fix

z16 committed:
Merge pull request #2160 from RubenatorX/patch-35

equipviewer validate bags fix

RubenatorX committed:
Wardrobe Fix

z16 committed:
Merge pull request #2161 from RubenatorX/patch-36

Wardrobe Fix

RubenatorX committed:
Gearswap Wardrobe inventories fix fix

RubenatorX committed:
Merge pull request #2163 from Windower/Gearswap-wardrobes-fix

Gearswap Wardrobe inventories fix fix

RubenatorX committed:
Gearswap Motes increase macro books

RubenatorX committed:
Merge pull request #2165 from Windower/gsmacrobooks

Gearswap Motes increase macro books

Chiaia committed:
Merge branch 'dev' of https://github.com/Windower/Lua into dev

Chiaia committed:
March 9th, 2022 Update

Byrth committed:
update digger messages for 2022-03

Byrth committed:
Merge pull request #2168 from Windower/update-digger-messages

update digger messages for 2022-03

z16 committed:
Merge pull request #2069 from trv6/pointwatch

pointwatch - automatically update message IDs

charitwo committed:
Update checkparam.lua

Add in "magic eva" to cover items like Nyame Sollerets which wasn't capturing magic evasion

z16 committed:
Merge pull request #2169 from charitwo/patch-1

Update checkparam.lua

posimagi committed:
[pointwatch] support Master Levels (specifically)

posimagi committed:
[pointwatch] add support for character level

posimagi committed:
[pointwatch] support character levels

posimagi committed:
Merge branch 'pointwatch-character-level' into pointwatch-master-level

posimagi committed:
update version number

posimagi committed:
fix syntax

z16 committed:
Merge pull request #2170 from posimagi/pointwatch-master-level

[pointwatch] support Master Levels (specifically)

posimagi committed:
[pointwatch] support job names and abbreviations

posimagi committed:
dynamically select client language

Byrth committed:
Merge pull request #2172 from posimagi/pointwatch-job-names-abbrs

[pointwatch] support job names and abbreviations

z16 committed:
Update pointwatch.lua

arshesney committed:
add yell and linkshell2 support

arshesney committed:
remove spurious comment

arshesney committed:
support for split linkshell and linkshell2

z16 committed:
Merge pull request #2173 from arshesney/blist-update

add yell and linkshell2 support

arshesney committed:
Fix weird interaction with gearswap

Byrth committed:
Merge pull request #2174 from arshesney/blist-fix

Fix weird interaction with Gearswap

alphaONE2 committed:
Add luasec library

Byrth committed:
Update digger messages for 2022-04 version

Byrth committed:
Merge pull request #2175 from Windower/update-digger-202204

Update digger messages for 2022-04 version

Byrth committed:
Update boxdestroyer for the 2022-04 version

Byrth committed:
Merge pull request #2176 from Windower/update-boxdestroyer-202204

Update boxdestroyer for the 2022-04 version

KenshiDRK committed:
[Slips lib] April 2022 Update

Adding the items from the April 2022 FFXI Update.

z16 committed:
Merge pull request #2178 from KenshiDRK/Slips

[Slips lib] April 2022 Update

Atsuke committed:
Support bug fix

Occasionally addon will try to resume crafting too quickly after interacting with support npc causing entire queue to be dumped with "you cannot craft at this time" errors.  Added a short delay in the check_queue function after poke_npc() call.

RubenatorX committed:
Allow for start/end pattern matching

RubenatorX committed:
fix print/whitespace

z16 committed:
Merge pull request #2179 from Atsuke/dev

Support bug fix

z16 committed:
Merge pull request #2074 from RubenatorX/patch-10

Update to properly handle more valid commands

z16 committed:
Merge pull request #2180 from Windower/RubenatorX-patch-1

FindAll -- Allow for start/end pattern matching

RubenatorX committed:
Pointwatch Fix broken Time Remaining

Byrth committed:
Make 0xDF and 0xE2 also update the alliance structure

Byrth committed:
Merge pull request #2183 from Windower/gearswap-expand-party-packet-parsing

GearSwap: Expand party packet parsing

funkworkz committed:
Update setbgm.lua

z16 committed:
Merge pull request #2185 from funkworkz/patch-4

Update setbgm.lua

RubenatorX committed:
Trusts - Re-Add Cornelia

Don't know why she was commented out in the first place

Byrth committed:
Add vana time offset

Byrth committed:
Merge pull request #2188 from Windower/adjust-duration-for-local-clock-drift

Add Vana Time offset

WinterSolstice8 committed:
[gearswap] Fix duration calculation in packet_parsing.lua

Co-authored-by: RubenatorX <RubenatorX@Gmail.com>

RubenatorX committed:
Merge pull request #2189 from WinterSolstice8/statuseffect_timestamp_fix

[gearswap] Fix buff duration calculation in packet_parsing.lua

KenshiDRK committed:
[Slips lib] Update for May 2022

Added items to slips libs for the 10 May, 2022 FFXI Update.

z16 committed:
Merge pull request #2190 from KenshiDRK/Slips

[Slips lib] Update for May 2022

Byrth committed:
get_player() sometimes returns nil when logging out

Byrth committed:
remove message printing from pointwatch

StarlitGhost committed:
[gametime] fix //gt route

I checked in Selbina and the arrival times are correct.

https://cdn.discordapp.com/attachments/484050756169760788/978709336333361222/unknown.png

StarlitGhost committed:
[gametime] bump version, add copyright date 2022

Nifim committed:
Merge pull request #2192 from StarlitGhost/patch-1

[gametime] fix //gt route

vessellaj committed:
Fix for player party member out-of-zone

z16 committed:
Merge pull request #2193 from vessellaj/patch-1

Trusts: Fix for player party member out-of-zone

RubenatorX committed:
DST time offset fix

z16 committed:
Fields: Refactored buff time calculation

z16 committed:
Fields: Incremented year count for buff time calculation

Byrth committed:
Add items for June 2022 Update

Byrth committed:
Merge pull request #2195 from Windower/update-slips-2022-06

Add items for June 2022 Update

Algelius committed:
Update Yush.lua

Added DIKs for f13-f15.

Algelius committed:
Update Yush.lua

Algelius committed:
Update Yush.lua

DIKs for f11 & f12

z16 committed:
Merge pull request #2196 from Algelius/dev

Yush.lua updated DIKs

Commit:
Added VerboseKeys setting.

Commit:
Wording update.

Commit:
Removed table identifier.

Commit:
parse_binds no longer reassigns key.

Commit:
output_text_table renamed, made list, and output text updated.

z16 committed:
Merge pull request #2197 from vessellaj/dev

Yush: Added a VerboseKeys setting to enable key display in each macro set.

RubenatorX committed:
[tables] add table.pack

RubenatorX committed:
[Gearswap][Motes-Mappings]  Ltng. Threnody

There is no spell called "Lightning Threnody"

KenshiDRK committed:
[fields lib] fix wrong 0x0E8 packet direction

0x0E8 is an outgoing packet.

Byrth committed:
Merge pull request #2201 from KenshiDRK/field-lib

[fields lib] fix wrong 0x0E8 packet direction

z16 committed:
Packets/Fields: Updated outgoing 0x061 length

KenshiDRK committed:
[Slips lib] Update for June 2022

Added items to slips libs for the 10 June, 2022 FFXI Update.

z16 committed:
Merge pull request #2205 from KenshiDRK/Slips

[Slips lib] Update for June 2022

Roland-J committed:
Adds the texts:draggable() function

Allows users to toggle the draggable flag post-creation. Previously, this flag could only be passed in with the config in a images.new(config).

z16 committed:
Merge pull request #2207 from Roland-J/texts.draggable

Adds the texts:draggable() function

Roland-J committed:
Implement image:register_event for click and drag

Adds the plumbing required for images to support image:register_event('left_click', function) and image:register_event('drag', function), potentially making images much more friendly to work with.

z16 commented on addons/libs/images.lua:
```diff
@@ -244,6 +246,15 @@ function images.pos_y(t, y)
+
+    last_pos.x = x
+    last_pos.y = y
+end
```
As discussed, this should be removed. The local `last_pos` variable should be used instead, similar to the `dragged` variable.

Roland-J replied:
How can the drag delta be calculated, from the addon's scope, without this helper function? My addon-side drag delta calculations fail without this, as my addon sees the  images.lua's last_pos var as a nil value; _addons appear unable to directly access last_pos_.

Please advise...
1) This bridges the gap, exposing it for addons to read. Is there a better way to bridge that gap?
2) How can we implement your requested change without breaking the ability for addons to calculate drag deltas?

z16 replied:
Addons should not calculate deltas. That is the lib's job. I'm still not convinced about the validity of this solution for your use case, but even in this case, the deltas could be sent as event parameters. I feel like this could be handled better in general, though. I'll have to think about this.

Roland-J replied:
AFAIK, when people are watching drag events, they are using it to keep objects in sync. In such a case, they are using a delta in one form or another.

**A. Top-Left Delta:** _(More complicated? Is this more reliable?)_
1. Find the current image's x/y distance from a bounding rectangle's top-left corner. (one that covers all the shapes)
2. Subtract x/y distance from the last-known position of the bounding rectangle to get the movement delta.

**B. Mouse-Movement Delta** _(Simpler? Is this as reliable?)_
1. Get the last known mouse x/y position
2. Subtract it from the current mouse x/y position to get the delta.

If drag events are going to be doing those calculations either way - _meaning the same overhead_ -  and getting the same result - _the same exact x/y delta, just calculated two different ways_ - it may be nice to provide the simpler method. Same overhead, but seems more user friendly for the addon authors, could be a nice service delivery for the dev base.

Passing it in as an event arg sounds convenient for the addon authors. Would that muck with anything else, though?

z16 replied:
> AFAIK, when people are watching drag events, they are using it to keep objects in sync.

You're literally the only one doing this, because this feature doesn't even exist yet :P

> Passing it in as an event arg sounds convenient for the addon authors. Would that muck with anything else, though?

No, but I still don't like it, this could be handled entirely in addon code. You'd have to keep track of the objects' original positions, then compare against the current, and adjust accordingly. Even with multiple directly draggable objects it wouldn't be too hard to do.

The reason this bothers me so much is because this feels out of scope for this lib, it's what a widget lib would be for. This lib only concerns individual objects, nothing outside of it, so your use-case of synchronizing with other elements feels like a weak justification.

As for your scenarios, only B makes sense. A is not more reliable. We only report discrete pixel values and they're fully accurate (to the pixel level).

Dev-Roland replied:
> > AFAIK, when people are watching drag events, they are using it to keep objects in sync.
> 
> You're literally the only one doing this, because this feature doesn't even exist yet :P

Why do you say that I'm the only one who syncs objects during drag? [As can be seen here,](https://discord.com/channels/338590234235371531/501099842097905675/994989166377050253) Akirane also does it, and there are likely others that do it as well that we're not aware of. It's not factual to say that I'm the only one, unless you're going on the technicality of _"using image:register_event('drag', fn)"_ vs _"windower.register_event('mouse', fn)"_, which would be missing the point if so.

As far as I can tell, if someone's watching drag then they'll be interested in the deltas - _that's what Akirane, probably unknown others, I are using the drag event for_ - as I don't yet see what else is important for an addon author during a drag.

> > Passing it in as an event arg sounds convenient for the addon authors. Would that muck with anything else, though?
> 
> No[...]

Great! Well, if the drag delta is the only important thing during a drag event, and we strip that delta from the event, we may be gutting the drag event from its usefulness.

> The reason this bothers me so much is because this feels out of scope for this lib, it's what a widget lib would be for. This lib only concerns individual objects, nothing outside of it, so your use-case of synchronizing with other elements feels like a weak justification.

Would a widget lib let the author bind (or whatever technical term) a bunch of images/texts into a cluster and then the widget lib would move them in tandem? That seems like a nice lib to have! I would agree that, if something like that is offered, that would seem like a better place for this. But, until then, what? The drag event could meet people's needs, if we build it to provide them the delta.

> We only report discrete pixel values and they're fully accurate (to the pixel level).

Great to hear!

Roland-J replied:
Akirane also does the same thing in one of their addons, as well as likely unknown others. However, this does seem like it would better belong in a _(not-yet-available)_ widgets lib, like you said. I guess for the meantime such authors should continue doing it themselves, right? Rather than shoehorning it in here, that is.

Roland-J commented on addons/libs/images.lua:
```diff
@@ -10,6 +10,8 @@ local meta = {}
 saved_images = {}
 local dragged
+local clicked
+local last_pos = {}
```
Declared this as a table right away, unlike dragged, so that line 415 can use the new function to assign the values without erroring out due to the function trying to declare keys on a var that was not yet declared as a table.

z16 replied:
Check `if clicked then` later on, this is a bug and only works because Lua is weakly typed.

z16 commented on addons/libs/images.lua:
```diff
@@ -407,6 +426,9 @@ windower.register_event('mouse', function(type, x, y, delta, blocked)
             if meta[dragged.image].root_settings then
                 config.save(meta[dragged.image].root_settings)
             end
+            if x == clicked.x and y == clicked.y then
```
I wonder if we need to be this exact with the click. Maybe a small tolerance is fine? Also maybe wrap it into a `settings.clickable` option?

Either way, this is currently under the `if dragged` condition, this should be moved outside in an analogous `if clicked` condition.

Roland-J replied:
Would an average delta of 2 be an ideal tolerance? Using my sample addon, I see that most of my gradual drags have an average delta of 0.5. 

I could change it to this:
`if (math.abs(x - clicked.x) + math.abs(y - clicked.y)) / 2 <= 2 then`
or (maybe safer)
`if (math.abs(x - clicked.x) + math.abs(y - clicked.y)) / 2 <= 1 then`

Please advise how you want the tolerance calculated. Thanks!

Roland-J replied:
Also, I placed it in the `if dragged` condition because all draggable objects would go into that if branch. I though that, since undraggable objects can't be moved, it'd be **unnecessary overhead** to make these undraggable objects get/set the clicked var.

Please advise if, following this explanation, you still find that this needs to change. Thanks!

z16 replied:
> Using my sample addon, I see that most of my gradual drags have an average delta of 0.5.

Huh? 0.5? How are you measuring delta, by pixel value? If so, an average may be the wrong tool. I'd use something like the 90th percentile. And even then I'd add something like a flat amount to that, to cover people who may be less accurate than you, in general.

There might be an overall better solution to this, but not sure. You seem to want to combine being draggable with being clickable, which makes this problem unnecessarily hard. Ideally I would solve those two problems separately and assume that being clickable and being draggable are mutually exclusive. The more I think about the more I think we should go that route. Would make this easy to solve - if `clicked` is not `nil` just check on mouse release if the mouse cursor is still within the text bounds, and if so, trigger the event. If you still want your behavior (which I really don't think you should, because it's user-unfriendly) you'd have to reimplement that part yourself.

> Also, I placed it in the if dragged condition because all draggable objects would go into that if branch.

This feature is unrelated to drag, even in your use-case, and so should be handled separately.

Roland-J replied:
> > Using my sample addon, I see that most of my gradual drags have an average delta of 0.5.
> 
> Huh? 0.5? How are you measuring delta, by pixel value? If so, an average may be the wrong tool. I'd use something like the 90th percentile. And even then I'd add something like a flat amount to that, to cover people who may be less accurate than you, in general.
> 

Yes, I am using pixel values. "_The mouse moved this many pixels on the x axis and this many pixels on the Y axis._" I'm not opposed to measuring it another way, and adding a flat amount sounds fine.

> There might be an overall better solution to this, but not sure. You seem to want to combine being draggable with being clickable, which makes this problem unnecessarily hard. Ideally I would solve those two problems separately and assume that being clickable and being draggable are mutually exclusive. The more I think about the more I think we should go that route. Would make this easy to solve - if `clicked` is not `nil` just check on mouse release if the mouse cursor is still within the text bounds, and if so, trigger the event. If you still want your behavior (which I really don't think you should, because it's user-unfriendly) you'd have to reimplement that part yourself.
> 

I see what you mean. When typing this comment, if I click down in the middle of the GitHub **Cancel** box, and then drag to the edge - _but still within the box_ - and release, it registers as clicking the box. This is a static setup, though.

On the other hand, browser favorites are similar to my implementation, they just have a healthy tolerance before moving from a click/release to a click/drag scenario.

https://user-images.githubusercontent.com/107378114/178127685-2f46e81d-0946-4605-80b3-2bef6a10d412.mp4

I can't say either method is better than the other. I think it's just a matter of taste/context.

Why don't we provide both? We could let draggable objects default to your suggestion - _no left_click events_ - but transition to the browser's left_click-if-under-threshold/drag-if-over-threshold behavior if the user adds a `image:clickable(true)` to an already draggable image. Viola, users can have it both ways.

For undraggable objects, see my next quote section. 

Thoughts?


> > Also, I placed it in the if dragged condition because all draggable objects would go into that if branch.
> 
> This feature is unrelated to drag, even in your use-case, and so should be handled separately.

Yeah, if it would be better that way, let's go for it! After seeing how GitHub does their cancel button - _click event fires on release if mouse is still anywhere in the button_ - I think I tried to oversimplify undraggable object clicks.

For these images, I think I should add code that verifies on unclick that the mouse is still within the bounds of the originally clicked button, and run the events only if it is. That way users can cancel the click event, by dragging out before releasing, incase the original clickdown was an accident/wrong button.

Sounds like a better experience, what do you think?

z16 replied:
> On the other hand, browser favorites are similar to my implementation, they just have a healthy tolerance before moving from a click/release to a click/drag scenario.

This is different because there are discrete places to move it to. It's not a continuous space you can move it around in, but only fixed places before and after other elements, so even massive tolerances before registering a click as a drag are fine. But in our case we either have to remove any tolerance, which would result in people occasionally moving a UI object accidentally instead of clicking it (which is not only annoying if it happens in an important situation, such as a battle, but it also destroys the UI layout, which, for people with certain levels of OCD, like me, is an even greater annoyance) or we make the tolerance big enough to tolerate cases like that, but make it significantly harder to fine-adjust UI elements (which, again, for people with certain OCD levels, like me, is a real annoyance).

> Why don't we provide both? We could let draggable objects default to your suggestion - no *left_click* events - but transition to the browser's left_click-if-under-threshold/drag-if-over-threshold behavior if the user adds a `image:clickable(true)` to an already draggable image. Viola, users can have it both ways.

I'm willing to let myself be convinced that this is useful - at this point I just don't see it. I know you seem to want it, though I have to say I still don't understand why, but I think we should ask people in the channel what they think. It would add noise to our API and I just don't see the benefit. Especially since this is something you could add on your own outside of the API.

That said, I'd be more willing to do it without the tolerance, and just let users mess up their UI if they're not careful. At least that way we don't have to have any special handling in our API and we can just blame the addon devs.

Dev-Roland replied:
Well, if we don't add the tolerance-before-dragging to (the new) draggable-and-clickable objects, addon authors won't be able to add that tolerance before dragging to their addons. That is because A) the images.lua is what is forcing all draggable objects to go into drag mode the moment the mouse is clicked down and B) requiring their userbase to modify their local images.luas to use their addons is likely a death wish for any hopes they had for users adopting their addon.

We would need to add that logic where, for the users who did `image:draggable(true)` and then `image:clickable(true)`, the drag event doesn't start on click until the mouse moves out of the tolerance zone. As is, the drag event starts immediately on click.

Dev-Roland replied:
If an author such as myself wanted to add this feature - _a tolerance before the drag event begins on draggable objects, where only a click event fires on release if the tolerance is not exceeded_ - would we be able to do that by packaging our own images.lua with our addon?

Roland-J replied:
If an author such as myself wanted to add this feature - _a tolerance before the drag event begins on draggable objects, where only a click event fires on release if the tolerance is not exceeded_ - would we be able to do that by packaging our own images.lua with our addon? I would think that would work.

What I'm wanting, like you said, seems to belong more in a new widgets.lua library, right?

Roland-J commented on addons/libs/images.lua:
```diff
@@ -244,6 +246,15 @@ function images.pos_y(t, y)
     t:pos(meta[t].settings.pos.x, y)
 end
 
+function images.last_pos(t, x, y)
```
Created this function in order to A) **(most importantly)** expose this local variable to addons and to B) imitate the get/set of the existing pos methods: _images.pos(), images.pos_x(), and images.pos_y()_.

z16 replied:
Addons should not be able to get that information at all outside of the event itself. Even then it's debatable.

Roland-J replied:
Drag events are probably going to be used to compute, and then act on, said delta. So it may be a great place to make it available.

z16 commented on addons/libs/images.lua:
```diff
@@ -407,6 +426,9 @@ windower.register_event('mouse', function(type, x, y, delta, blocked)
                 config.save(meta[dragged.image].root_settings)
             end
+            if x == clicked.x and y == clicked.y then
+                call_events(dragged.image, 'left_click')
```
I was debating whether we should leave it as `left_click` or just call it `click` and have `right_click` as the odd one out, because it's used more rarely. I'm not sure though, would be up for discussion. Could also alias it so `click` and `left_click` are identical in behavior.

Roland-J replied:
I should be able to easily add support for right click, too. Would it be silly to assume that type 3 = right click and type 4 = right release? I would prefer adding that functionality while we're at it, for the community's sake.

1. Please advise the types (integer) for the right click/release events. (I don't see them documented anywhere.)
2. Please advise how I would setup the alias, if you decide that would be desired. Thanks!

z16 replied:
> Please advise the types (integer) for the right click/release events. (I don't see them documented anywhere.)

No idea, easy for you to test. Just `windower.register_event('mouse', print)`, then right click. For reference, it's the `wParam` parameter of [`MouseProc`](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms644988(v=vs.85)), `& 0xFF`. Whatever that is.

> Please advise how I would setup the alias, if you decide that would be desired.

Actually, I'm tending towards leaving it as `left_click` here, but also handling `click` by looking up a mapping table when registering it. Can define a table like this:
```lua
local event_map = {
    click = 'left_click',
    double_click = 'double_left_click',
}
```

Then use this when registering the event, at the very top of the function:
```lua
    key = event_map[key] or key
```

z16 commented on addons/libs/images.lua:
```diff
@@ -407,6 +426,9 @@ windower.register_event('mouse', function(type, x, y, delta, blocked)
+            if x == clicked.x and y == clicked.y then
+                call_events(dragged.image, 'left_click')
+            end
             dragged = nil
```
Like `dragged` before, `clicked` should be unset here.

Roland-J replied:
Thanks, done!

Roland-J commented on addons/libs/images.lua:
```diff
@@ -387,17 +398,25 @@ windower.register_event('mouse', function(type, x, y, delta, blocked)
         if dragged then
             dragged.image:pos(x - dragged.x, y - dragged.y)
+            call_events(dragged.image, 'drag')
+            dragged.image:last_pos(dragged.image:pos())
```
This needs to be called after the call_events, not before, in order for the event to retain access to the previous last_pos. Update this after the event so that the next drag event has an updated last_pos.

Roland-J commented on addons/libs/images.lua:
```diff
@@ -407,6 +426,9 @@ windower.register_event('mouse', function(type, x, y, delta, blocked)
             if meta[dragged.image].root_settings then
                 config.save(meta[dragged.image].root_settings)
             end
+            if x == clicked.x and y == clicked.y then
```
We are only registering left_click events for draggable images if the image wasn't dragged between click and release.

Roland-J commented on addons/libs/images.lua:
```diff
@@ -387,17 +398,25 @@ windower.register_event('mouse', function(type, x, y, delta, blocked)
 
     -- Mouse left click
     elseif type == 1 then
         for _, t in pairs(saved_images) do
```
**SEPARATE PR:** If we wanted to, this could be changed to a backwards iteration that, on the first match, does all the existing stuff but also breaks out. This may be useful for addons that have stacked images, to make the click event only fire for the addon's top-most image in the mouse's x/y region.

Roland-J reviewed (commented):
Added some clarifying comments

Roland-J committed:
Unset clicked on left release

Sets the clicked var to nil on left mouse release

Roland-J commented:
@z16 Thank you for the feedback! I am trying to implement the changes but need some clarifications. Please see my follow up questions. Thanks!

Roland-J commented:
@z16 Sorry for the delay, but I've implemented much of what was mentioned, plus a lot more. It can be found and demo'd in a fully functional form here: _(feel free to re-use this code)_
https://github.com/Roland-J/SmartSkillup/blob/main/libs/sms_images.lua
https://github.com/Roland-J/SmartSkillup/blob/main/libs/sms_texts.lua

It's a lot of changes - _trying to flesh out the event self-registration feature as much as possible to lower the barrier to UI development and encourage more UI development_ - so there may be too many conflicts to event attempt it in the PR, so I figured I'd first mention it before merging it in. I'll take my next action according to the feedback.

Summary:
1. Added hover/unhover event. _(Only fires if said event has been registered for said element)_
2. Added scroll up/down events. _(Only fires if said event(s) have been registered ffor said element)_
3. Added scroll click event. _(Only fires if said event has been registered for said element)_
4. Added right click/release event. _(Only fires if said event has been registered for said element)_
5. Added right drag.  _(Only occurs if its new flag is enabled.)_
6. Added right drag event. _(Only fires if said event has been registered for said element)_
7. Added scroll drag.  _(Only occurs if its new flag is enabled.)_
8. Added scroll drag event. _(Only fires if said event has been registered for said element)_
9. Differentiated left drag from right drag. (Renamed vars & func and provided alias vars and funcs for legacy support)
10. Finished the drag tolerance feature. _(Defaults to 0, drags only start once tolerance is exceeded)_
11. Refined event sinking with event checking - keep going to first event-registered element - to allow events to sink through a irrelevant hitbox or similar item into the topmost relevant item.
12. Added event returning - _return out on first relevant element_ - to avoid allowing for both the intended (topmost) element and unintended (lower) elements to trigger events; only the intended (topmost) one will trigger.

I've added some event variables to them. You can see what I'm doing with those variables in https://github.com/Roland-J/SmartSkillup/blob/main/libs/sms_ui.lua

Nifim commented:
Can you update this PR? we had a change to our repo that caused discontinuity between the various forks.

Once you do I can review your code and get this back on track to being merged.
